### DISCUSSION TITLE
GA, Amplitude util 함수의 방향성에 관하여

### DISCUSSION THREAD

**Discussion by p-iknow** [EveryAnalytics/react-analytics-provider#141] [Ideas]
### 해당 논의의 배경

- https://github.com/EveryAnalytics/react-analytics-provider/discussions/127#discussion-3533587  (googleAnalytics.click이 있다면 amplitude.click 도 있어야 할까?)
- ^ 위 discussion에서 부터 아래 제시한 방향성이 도출되었습니다. 위 링크에 섞이면 논의 주제가 주목되지 못할 것 같아 새로운 discussion을 만들었습니다. 

### 문제제기
googleAnalytics.click은 googleAnalytics.event를 우리식대로 한번 감싼것. '우리식대로 감싼'게 오히려 사용자들에게 혼란을 주지 않을까? 그리고 이 설계를 통일하기 위해 amplitude.click 개념도 우리식대로 만들면 더 혼란스럽지 않을까?

### **제안하는 방향성**

- GA, Amplitude 각 라이브러리 기본 API에 대한 헬퍼만 제공하자.
- e.g.
    - 기본
        - googleAnalytics.event(...) :  ga의 initialize스크립트를 쉽게 쓸 수 있는 헬퍼
        - googleAnalytics.event(...) // :man-gesturing-ok: ga의 event란 api를 쉽게 쓸 수 있는 헬퍼
    - 기본이 아닌 헬퍼
        - googleAnalytics.click(...): ga는 click이란 api가 없다.
        - googleAnalytics.modal(...): ga는 modal이란 api가 없다.

### 이유

- 지금 만들고 있는 util 함수들을 너무 공급자 위주로 만들고 있다. util 함수들은 사용자의 필요에 의해서 만들어져야 한다. 우리가 만들고 있는 util 함수들이 얼마나 쓰일지, 얼마나 필요한지 우리는 잘 모른채 만든다.
- 우리가 만든 ga 나 amplitude util 함수의 기본적인 기능만 있어도 react-analytics-provider를 쓰는데 큰 지장이 없다.
- 아주 기본적인 util를 제공한 뒤 beta open을 하고, 사용자로 부터 피드백을 얻어 개발 방향성을 얻어 가는 방식이 더 에자일 하다.
- util 함수를 많이 만들면 그 만큼 api 에 대한 문서화 등의 유지보수가 많이 들어간다. 초기에는 기본적인 util만 만들고, 관련 api 문서를 ga, amplitude 원래 문서에 위임하면, 유지보수 비용이 덜 들어간다.

### **방향성에 공감한 뒤 정해야할 것**

- 무엇이 기본적인 util인가, GA, Amplitude 각자가 다른 api를 제공하므로 기본적인 util은 GA, Amplitude에서 각자 정한다.
- 기존에 만들어둔 함수들이 위 기준(기본적인)에 해당하지 않으면 잠정적 삭제

**Comment by jhlee910609** [comment]
@p-iknow 좋은 아젠다라고 생각합니다! 👍 우선 저는 각각의 Product Analytics tool들의 공식 문서 기준으로 basic한 util 함수만 제공해도 괜찮지 않을까 싶습니다. 이외의 이벤트들은 사용자에게 event 기본 함수만 제공해주는 게 좋을 것 같습니다. GA를 예시로 들면 아래 이미지와 같이 event, pageView, timing, exception 와 같은 이벤트 함수들만 만드는 게 조금 더 혼돈을 줄일 수 있을 것 같습니다.

<img width="265" alt="스크린샷 2021-08-25 오후 10 37 27" src="https://user-images.githubusercontent.com/28521282/130800682-25ce0a51-c0bb-49bf-89c7-f6439fcc6aa1.png">

util 함수의 네이밍도 당연히 공식 문서 기준으로 맞추는 게 좋을 것 같고요. 그리고 기본 util의 정의는 각각의  Product Analytics tool에 따라 정해지는 게 좋을 것 같습니다. GA는 GA대로, Amplitude는 Amplitude대로요~

**Comment by greatSumini** [comment]
@p-iknow 오 정리 대박이네요.. 감사합니다 ㅎㅎ
사실 저희가 이슈 메이킹 단계에서는 엄청 많은 고민을 거치지는 않고 있다보니 방향성이 약간 엇나갈 수 있었는데, 이렇게 잘 집어주셔서 다시 한번 돌아볼 수 있는 기회가 된 것 같습니다 😊 

서비스 런칭 템포를 빠르게 가져갈 수 있다는 점에서 정말 좋은 아이디어인 것 같습니다!

다만 원래의 방향성이 옳았는가와 무관하게 'Provider와 util 함수의 기본적인 기능만으로도 Minimum Viable Product인가'라는 점에서는 조금 걱정이 됩니다. 제가 RAP 릴리즈하자마자 바로 쓰고싶어서 욕심내는 것일 수도 있습니다 ㅎㅎ 생각해보니 이건 별도의 문제인 것 같기도 하네유

**Reply by milooy** [reply]
릴리즈하자마자 바로 쓰실 때 
없어서 아쉬운 함수가 모가 있을까요? 👀 
예시 들어주심 넘 도움되겠어요 ㅎㅎ

**Reply by greatSumini** [reply]
@milooy 현재도 거의 event만 사용하고 있었어서 아쉬운 함수는 없을 것 같아요!
다만 기존의 익숙한 구조(react-ga또는 직접 만든)를 벗어나 새로 공부할만큼, 또는 기존 프로젝트를 전환할만큼의 메리트를 느끼진 않을 것 같습니다. 이유는 다음과 같습니다.
1. 저는 ga만 사용해서 😅 앰플리튜드 지원의 이점을 누리지 못 합니다.
2. 저는 작성하는 ga 관련 코드의 대다수가 각 이벤트들에 대한 wrapper 함수인데, 기본 기능만 지원한다면 RAP를 써도 이 코드들은 그대로 작성해줘야합니다.
3. react-ga와 다르게 Context로 App을 통째로 감싸는데, 위의 두가지 이유와 맞물려 기대한만큼의 편의성이 없는 것 같습니다 😭  확실한 리턴이 없다면 Context 도입은 꺼릴 것 같습니다. (AxiosProvider로 App을 통째로 감싸고 axiosContext.request를 호출하는듯한 느낌..?)

물론 미니 프로젝트에 가볍게 적용해보는등 개밥먹기는 하겠지만, 밥으로서 먹진 않을 것 같습니다..!

**Reply by uoon-dev** [reply]
저도 @greatSumini 의견에 공감해요. 기본적인 기능만 구현해서 라이브러리에서 제공해주는 편의가 줄어든다면 굳이 서비스에 도입하려는 리스크를 감수하진 않을 것 같아요. 많은 래퍼 함수를 만들어서 라이브러리를 사용하고 싶은 메리트를 살리는 게 영업(?)을 위해서도 중요하다 생각합니다 :)

**Reply by milooy** [reply]
공감이 됩니다.
수민님이 말씀해주신 니즈는 context 없이 googleAnalyticsHelper만 써도 충족이 될거같다는 생각이 드네요!

이런 방법은 어떨까요?
- @every-analytics/react-analytics-provider/core
- @every-analytics/react-analytics-provider/google-analytics-helper
- @every-analytics/react-analytics-provider/amplitude-helper

로 라이브러리를 쪼개면,
core에서 아예 분리되었으니 google-analytics-helper와 amplitude-helper 는 각 Product Analytics툴 각각의 특성을 더 살리는 방향성으로 가도 괜찮지 않을까요?

GA만 쓰는 사람은 google-analytics-helper 만 쓰다가,
- 반복되는 코드 줄이고싶다
- 로깅 컴포넌트 사용하고 싶다 (`AnalyticsClick` 등)
- GA외에 다른 Product Analytics 툴도 도입하게 되었다
등의 니즈가 생기면 context가 있는 core도 쓸 수 있도록!

**Reply by sensecodevalue** [reply]
@milooy 님의 방식대로 한다면 GA와 Amplitude 각각 현명하게 환경에 맞게 사용할 수 있을 것 같아요!

저도 여러 @greatSumini @uoon-dev 님의 말씀처럼 여러 함수를 Wrapper해서 편의성을 증진시키는 방법에 대해서 찬성입니다.
GA에서는 여러 추천 이벤트들이 많더라구요!

공용으로 많이 쓰이는 것들: https://support.google.com/analytics/answer/9267735
온라인 스토어 : https://support.google.com/analytics/answer/9268036
게임 : https://support.google.com/analytics/answer/9267565
...

**Comment by milooy** [comment]
@jhlee910609 GA API 정리 감사합니다~

앰플리튜드의 기본 API 는 다음과 같은데,
1. Initialization: `init`
2. Setting User Id: `setUserId`
3. Sending Events: `logEvent` 
4. User Properties: `identify`, `setUserProperties`

기본 API자체가 워~낙 깔끔해서 헬퍼가 필요 없을 정도네요 🤔 
https://developers.amplitude.com/docs/javascript

현재 만들어진 'amplitude instance 반환 함수' 만 있어도 충분할듯... ㅎㅎㅎ;;

**Reply by jhlee910609** [reply]
api가 깔끔하네요 😀
